diff --git a/ranger/core/main.py b/ranger/core/main.py
index 598ce243..ade8fd3d 100644
--- a/ranger/core/main.py
+++ b/ranger/core/main.py
@@ -435,6 +435,12 @@ def load_settings(  # pylint: disable=too-many-locals,too-many-branches,too-many
                     fm.notify(ex_msg, bad=True)
             ranger.fm = None
 
+        ranger.fm = fm;
+        for plugin in __plugins__:
+            import importlib
+            fm.commands.load_commands_from_module(importlib.import_module(plugin))
+        ranger.fm = None
+
         allow_access_to_confdir(ranger.args.confdir, False)
         # Load rc.conf
         custom_conf = fm.confpath('rc.conf')
diff --git a/ranger/plugins/__init__.py b/ranger/plugins/__init__.py
new file mode 100644
index 00000000..e69de29b
diff --git a/setup.py b/setup.py
index edf48c4a..ffd76b99 100755
--- a/setup.py
+++ b/setup.py
@@ -139,6 +139,7 @@ def main():
             'ranger.ext.vcs',
             'ranger.gui',
             'ranger.gui.widgets',
+            'ranger.plugins',
         ),
     )
 
